src/main/flight/pid.c

```diff
@@ -545,27 +542,23 @@ void pidInitConfig(const pidProfile_t *pidProfile)
 #endif
 
 #ifdef USE_DYN_LPF
-    if (pidProfile->dyn_lpf_dterm_idle > 0 && pidProfile->dyn_lpf_dterm_max_hz > 0) {
-        if (pidProfile->dterm_lowpass_hz > 0 ) {
-            dynLpfMin = pidProfile->dterm_lowpass_hz;
-            switch (pidProfile->dterm_filter_type) {
-            case FILTER_PT1:
-                dynLpfFilter = DYN_LPF_PT1;
-                break;
-            case FILTER_BIQUAD:
-                dynLpfFilter = DYN_LPF_BIQUAD;
-                break;
-            default:
-                dynLpfFilter = DYN_LPF_NONE;
-                break;
-            }
+    if (pidProfile->dterm_lowpass_hz > 0 && pidProfile->dyn_lpf_dterm_max_hz > pidProfile->dterm_lowpass_hz) {
+        switch (pidProfile->dterm_filter_type) {
+        case FILTER_PT1:
+            dynLpfFilter = DYN_LPF_PT1;
+            break;
+        case FILTER_BIQUAD:
+            dynLpfFilter = DYN_LPF_BIQUAD;
+            break;
+        default:
+            dynLpfFilter = DYN_LPF_NONE;
+            break;
         }
     } else {
         dynLpfFilter = DYN_LPF_NONE;
     }
-    dynLpfIdle = pidProfile->dyn_lpf_dterm_idle / 100.0f;
-    dynLpfIdlePoint = (dynLpfIdle - (dynLpfIdle * dynLpfIdle * dynLpfIdle) / 3.0f) * 1.5f;
-    dynLpfInvIdlePointScaled = 1 / (1 - dynLpfIdlePoint) * (pidProfile->dyn_lpf_dterm_max_hz - dynLpfMin);
+    dynLpfMin = pidProfile->dterm_lowpass_hz;
```

Comment:
I am not sure that reusing `pidProfile->dterm_lowpass_hz` to set the dynamic LPF minimum frequency is the right thing to do here, in the interest of having a clean and not intermingled configuration, and to allow users to experiment with different settings.
What do others think?

Reply:
I see it correct because enabling the dyn lpf it disables the static lpf. If we use two different variables maybe the user can think he has two different filters and that is not true.

Reply:
Ok. But the way one parameter is now used for two different things is very confusing. It sould be used as the dynamic lowpass mid frequency (i.e. the equivalent of the static frequency), or renamed to something like `dterm_lowpass_static_mid_or_dyn_min_hz` to make it indicative of the different ways it is used.

Reply:
I don't know too much about filters, but I think it will be the mid frequency if it was a notch. It is a lpf, so it will be the lower cutoff.
So to me, this parameter is always the lower cutoff, the only different is that if it the user configures it dynamic, this lower cutoff will go upper with the throttle, if not, it will remain static, but in both cases the min cutoff will remain the same.

Reply:
@McGiverGim: Common sense would suggest that if you take a static setting (like the lowpass cutoff) and switch it to dynamic, if you were to reuse the old parameter that had been used to indicate the static cutoff, this would be used for the middle of the the range of values that the dynamic setting can take on, not at one arbitrary end of the value range.

Reply:
The purpose with dynamic lowpass is not to allow a lower minimum than before.  

It is to give the user the chance to move the lowpass value even higher than their usual value as they increase throttle. 

Our test experience was  that the minimum value, in dynamic mode, would typically end up being close to, or a bit higher than, the static value the user found worked well previously.  However we know that we can go higher as throttle increases.  

Hence the correct starting point for the dynamic minimum, with dynamic lowpass, would be the user's previous lowpass1 value.  

From a defaults perspective, our recommended default would anyway have been almost the same number as the previous lowpass1 default.  Both should be something like 150Hz going forward - in my opinion, anyway.

Reply:
@ctzsnooze: The fact that it takes you 4 paragraphs to explain why you think that these two settings should be crammed into one is a good indicator that what you are trying to do is going to be confusing for the user who only sees the name of the setting, and not the 4 paragraphs of explanation, when trying to change settings in the field.
So if this is really the way to go, then we should rename this to something more indicative of what it is used for, like `dterm_lowpass_fixed_or_dyn_min_hz`, and not just leave it cryptic.

Reply:
@ctzsnooze:

> If we rename this value to 'dyn' or 'min' anything, but the user has not enabled dynamic lowpass filtering, then the name isn't correct, because the cutoff is static all the time. If we don't include 'dyn' or 'min' in the name, then when dynamic is active, the label isn't correct either.

That's exactly my point. We are trying to optimise a couple of bytes here for by sacrificing accuracy of naming - not sure if this is worth it.

Reply:
@ctzsnooze:

gyro_lowpass_hz => **minimum** frequency
gyro_lowpass_max_hz => maximum frequency

Just looking at this tells me that the logic in this is utterly flawed.

Reply:
@mikeller:

gyro_lowpass_hz => same old gyro_lowpass_hz as always
gyro_lowpass_max_hz => what it can go up to, if you want it to

:-)

Reply:
@ctzsnooze:

> gyro_lowpass_hz => same old gyro_lowpass_hz as always

That's what I asked to be done right from the start - you are the one who is proposing to make this into an ugly and confusing mess by using `gyro_lowpass_hz` for things it isn't.

Please stop making stupid comments, if you don't want this merged say so and we can close it.

Reply:
We agree that it is easy to add gyro_lowpass_min_hz as a separate entity.  We disagree on whether it is necessary.  Our respective positions are clear.  Let's see what other pilots / devs think.  @etracer65 @McGiverGim @wind0r ?

Reply:
That's a great idea! 

So for say the gyro lowpass 1 we can have options PT1, BIQUAD, and DYN_BIQUAD say, and likewise for D?

Default can remain the same PT1 as before I guess, if user selects BIQUAD they would have had different results and probably needed to change the cutoff anyway, now they can choose DYN_BIQUAD if they wish.

I'd be very happy with this solution.  Awesome.

Reply:
And add dedicated min and max parameters for the dynamic mode leaving the existing lowpass cutoff to only relate to the static mode. There will need to be some sanity checked by in config.c to make sure the invalid config aren’t selected.

Reply:
Let's say the user can choose, for lowpass 1 mode, either PT1, BIQUAD, or DYN_BIQUAD.

And lets say the default lowpass 1 mode remains PT1 and the value remains the same.

Now if the user was to change from PT1 to Biquad, they would need to change their lowpass value, since the behaviour is different.  When they make the conscious step to customise the behaviour in that way, they need to know how Biquad is different from PT1, and then should customise the value appropriately.

We do not currently have a separate default lowpass 1 value for PT1 mode vs Biquad mode.

As I see it, providing dynamic lifting of the lowpass 1 value through a mode selection is much the same.  If the user selects 'DYN_BIQUAD' instead of PT1, it is not unreasonable to expect that they might need to modify the value - just as if they had selected 'BIQUAD'.

As it happens, the default value for PT1 is perfectly acceptable as the default for DYN_BIQUAD.

Further, certainly for testing, it is much easier for LUA or OSD users if the current lowpass value applies to DYN_BIQUAD when that mode is selected.  They can then test and tune it on the field.  

Re-using lowpass 1 hz in this manner avoids *yet another* entry not only in CLI but also LUA and OSD.  

The same default values are appropriate for PT1 lowpass and DYN_BIQUAD min.  If you flew the two you'd find that the transition from one to the other, with the same values, works very acceptably.  

I honestly see no positives at all in providing an independent minimum for dynamic lowpass, and I see lots of negatives, the biggest of which is that it would seriously constrain field testing.

Reply:
@ctzsnooze: If you don't see positives in not introducing a mangled and messed up configuration that is your prerogative. You also do not have a stake in keeping things like configurator and the lua scripts maintained, which will become a lot harder if we do not stick to using clean configuration settings.
So you got two choices now: Do this cleanly, or don't do it at all.

Reply:
I am also in favor of introducing new parameters / lowpass type options, specifically for this concern and/or not re-using previous parameters.

The general concept in software engineering is _Separation of Concerns_ or _Single Responsibility Principle_ (and related patterns).

In general, the _responsibility_ is a 'reason to change'. Imagine the filtering subsystems are changed for some other reason (e.g. refactoring, rewrite, remove, or rename). There are now _multiple_ discrete concerns for individual parameters - one potentially substantive, one potentially cosmetic (or similar). 

These are actually separate responsibilities, and as we have no suitable construct or mechanism in the language that the firmware is written in, we should err towards separate, isolated parameters - temporarily suffering the duplication - until the old interfaces are deprecated and removed, if ever.

Such a design also naturally supports Backwards Compatibility while mitigating escaped defects (e.g. from introduced multiple concerns / responsibilities).

For field testing I would suggest adding a new (or alternate) CMS Menu which exposes the Dynamic Lowpass / Cascaded Notch related functionality, as Launch Control did - it's otherwise a tangential concern.

Reply:
What I propose minimises the changes that would be needed in Lua and OSD.  Since I use LUA a lot, I do have a stake in any code structure that weakens its usability for me.

The current PT1 / BIQUAD options for the gyro lowpass carries the same fundamental problem that worries you here.  Both those options, despite very different results, share the same value. 

If we were to follow your logic, the settings should be like this:

gyro_lowpass_mode (PT1, BIQUAD, DYNAMIC)
gyro_lowpass_pt1_hz
gyro_lowpass_biquad_hz
gyro_lowpass_dynamic_min_hz
gyro_lowpass_dynamic_max_hz

Then the user could change mode and have 'perfect' and different values stored for each mode.  I think, in principle, that is what you are suggesting should be done.  I have some sympathy for this; its an elegant, perfect solution.

My perception, however, is that pilots just don't care about this level of perfection.  They do care about clutter in the UI, the ever growing footprint causing features dropping off some targets, and the huge list of settings where its tough knowing which setting does what.

The fact is that pilots have NEVER complained about how changing mode between PT1 and BIQUAD gives very different filtering results, despite sharing the same value, speaks volumes for the fact that most pilots don't have a problem with the concept of one CLI value achieving different results in different modes. 

That's why after logical and unemotional reflection, and after giving due weight to your concerns, I still advocate the simpler:

gyro_lowpass_mode (PT1, BIQUAD, DYNAMIC)
gyro_lowpass_hz
gyro_lowpass_dynamic_max_hz

The more we clutter up betaflight with unnecessary CLI settings the more difficult we make it for everyone.  This keeps it simple and makes it easier to test right now without any LUA or OSD changes.

Reply:
@etracer65 @fujin could you please review the above?  I think the optimal solution is perhaps a little different now that the dynamic option goes into the lowpass mode user selection.

Reply:
Thanks @fujin. One small point:

> Finally, I will concede that it can be beneficial to learn such principles by safely attempting, and failing, then being guided through the "why"

Unfortunately, where this discussion is concerned, we are talking about a range of options that impacts on how a published API will be defined - an API that includes provisions for retaining backwards compatibility, in order to allow pilots a certain amount of leniency when updating the firmware on their craft. As a consequence, this means that the cost of 'failing' includes having the failure set in stone for a long time after the API has been released, even after the failure been recognised and fixed, by having to support the failed version of the API with workarounds / special cases in every client that uses the API.

Betaflight has failed in this respect in the past, and learned this the hard way (if you want to see proof of this look into the special cases that are handled in https://github.com/betaflight/betaflight-configurator/blob/master/src/js/msp/MSPHelper.js), and one of the learnings is that it is worthwhile to use scrutiny, caution, and best practices for everything that affects the MSP API, as any error there will be costly for a long time after it was made.

Reply:
The general logic of presets is well explained above, and I thank you for it.  The response does not however identify what safety or quality risk exists in the simpler proposal that is being addressed by the more complex proposal.

Without any explicit detail as to what you are actually concerned about in this specific instance, the generalities are guidelines only, and the case should be determined on the merits or otherwise of the alternate suggestions.

In my estimation, the selection between "PT1, BIQUAD, or DYNAMIC" for lowpass is a "single concern" choice.  All influence filter behaviour and how the quad behaves.  That the cutoff remains the same actually is a minor point; the user changes them to change behaviour of the quad, to counter noise; each choice results in some different behaviour.  So I think when we introduced FKF we put it here in the same way.  I don't see any issue with putting DYNAMIC here, nor do I think anyone does.  

The issue at hand is whether using the gyro_lowpass_hz value for the static value which applies as the lowpass for a fair bit of the lower end of the throttle curve is a hazard.

If the previous PT1 value was not hazardous, and the user selected DYNAMIC, it would not be the lowpass value that would present a hazard, any more than if they selected BIQUAD.

So I fail to understand what the safety hazard is that you are concerned by, and to date NO-ONE has explained what they think the safety hazard is.

Of course there IS a safety hazard with choosing DYNAMIC, but that arises from the MAX value, which could make the motors hot.  Not the starting value, which either would be the default (fine, no dangerous than choosing biquad), or what the user had previously customised it to (also safe since cutoff does not change).

So please explain the safety hazard, because I just don't get it.

This for me is a practical matter, not a theoretical or emotional matter.  The question is practical: what exactly do you perceive to be the actual safety hazard.  What exactly, in this specific case, do you think people might do that would achieve something bad?

Because for anything to happen, the user has to select 'DYNAMIC', they have to think about a default MAX; if they do this without reading anything at all it will be perfectly safe (as far as the concept is safe and the max is safe; there is no problem at the low end, since the cutoff is the same in dynamic at the low end).

I am unable to understand why you cannot identify a safety hazard and in the absence of ANY single clear identifiable concern I restate that I do not agree with what you propose.

None the less it is apparent that I must concede; I give up.

Just bear in mind that, consequently, no-one can easily change their initial cutoff value in dynamic lowpass mode via their OSD or LUA, until LUA and OSD get updated.  For a parameter that we would value field feedback on, this is a significant inconvenience.  Lua already has issues with out of memory for number of parameters, just adding one more won't help.

Reply:
Hey, I'm sorry - you kind of honed in on what I said regarding safety as "hazardous" which is not where I was going at all. I was referring to safety from (our) the maintainer's perspective regarding the software's behavior - software backwards compatibility, and understanding for future changes during routine work such as refactoring, or testing. I agree though, the user must be safe while operating the system, including motor heat and such. I believe you have picked an adequate maximum hz (whatever the lowest level interface name becomes), it worked well when I tested it. I have no objection there.

Regarding outer layer interfaces and naming, if a new parameter is introduced it won't be added to public MSP protocol for a little while. As I mentioned earlier, this decoupling allows for simplification of work on a given layer's module or function.

I do not know that any potential MSP protocol changes should be included in 4.0.0 release, although Semantic Versioning 2.0 would allow it as part of the major breaking change release - @mikeller? We would need time to agree on naming to stabilize the proposed MSP protocol changes first.

That means you would have to use custom Lua scripts from a branch if you wanted to work with new MSP protocol (even with 4.0.0), and potentially a custom configurator from a branch, until 4.1.0 like we would normally do. This gives us time to experiment with the settings and CMS interfaces without having to port all of those changes to the MSP based interfaces which is strictly versioned which is the primary reasoning for the decoupling. During development, the CLI interface settings (and parameter groups) and the CMS (OSD) interface are fastest to change, with minimal dependencies (maybe Blackbox?)

I think the case for re-using temporarily gyro_lowpass_hz is an interesting but fundamental problem from a maintenance perspective. I will agree it worked well in my exposure to it during your team's development of the dynamic lowpass and cascaded notches which I commend.

Doing version-specific behavior in the outer interfaces (MSP, Lua, Configurator) is *very* bad (should any arise), and we would be stuck with it (as well as supporting the old behavior), if that is where the control for the dynamic lowpass is done.

So, I still believe that dynamic specific parameters should be introduced. I would also do the OSD CMS menu work for the new parameter, I can help with this. I can also show you how to do the MSP part and a custom Lua script which can be distributed and changed more frequently, separate from the normal Betaflight release life-cycle.

Reply:
@fujin: The Betaflight release version number and the MSP API version number are pretty much independent, and have been treated like this in the past. This means that Betaflight 4.0.0 does not give us free reign to introduce breaking change into MSP. I do not think that it would be advisable for us to do so at any time, as this would either be massively disruptive for our users, or else require us to implement and maintain two independent  MSP frontends in configurator, putting a huge strain on our resources for little benefit.

As for @ctzsnooze's concern about making the parameters available for users to tune: For rapid testing / feedback this should be done in CMS - this is the one interface where we can move quickly without being encumbered in API / backwards compatibility constraints.
Doing it in lua won't work well either way - if we use the approach of hacking `gyro_lowpass_hz` to apply to `gyro_lowpass_min_hz` then users will only be able to change the minimum frequency in the field, but not disable the dynamic filter or set the maximum frequency. To me this sounds even more incomplete and confusing than not having any of the settings available in lua for now.

Re @ctzsnooze's concern about lua memory overflow: Since each page of the script is loaded independently, and unloaded before the next page is loaded, adding more settings to a page will only cause memory issues if there are too many settings on this page already, and in this case it can be easily mitigated by splitting this page into two separate pages - which is normally needed anyway for QX7 / X9D, since the display size cannot really hold more elements than will fit into memory.

Re lua and versioning, I have tinkered some, and come up with https://github.com/betaflight/betaflight-tx-lua-scripts/pull/188 that allows lua pages to be loaded dependent on the MSP API version.